pallets/stableswap/src/lib.rs

```diff
@@ -712,14 +774,53 @@ pub mod pallet {
 				Error::<T>::InsufficientBalance
 			);
 
-			let (amount_out, fee_amount) = Self::calculate_out_amount(pool_id, asset_in, asset_out, amount_in)?;
+			let pool = Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+			let pool_account = Self::pool_account(pool_id);
+			let initial_reserves = pool.balances::<T>(&pool_account).ok_or(Error::<T>::UnknownDecimals)?;
 
+			let (amount_out, fee_amount) = Self::calculate_out_amount(pool_id, asset_in, asset_out, amount_in)?;
 			ensure!(amount_out >= min_buy_amount, Error::<T>::BuyLimitNotReached);
 
-			let pool_account = Self::pool_account(pool_id);
 			T::Currency::transfer(asset_in, &who, &pool_account, amount_in)?;
 			T::Currency::transfer(asset_out, &pool_account, &who, amount_out)?;
 
+			let share_issuance = T::Currency::total_issuance(pool_id);
+			let assets = pool.assets.clone();
+
+			let state = PoolState {
+				assets: pool.assets.into_inner(),
+				before: initial_reserves.iter().map(|v| v.into()).collect(),
+				after: initial_reserves
+					.into_iter()
+					.zip(assets.iter())
+					.map(|(v, asset_id)| {
+						if *asset_id == asset_in {
+							v.amount.saturating_add(amount_in)
+						} else if *asset_id == asset_out {
+							v.amount.saturating_sub(amount_out)
+						} else {
+							v.into()
+						}
+					})
+					.collect(),
+				delta: assets
+					.into_iter()
+					.map(|v| {
+						if v == asset_in {
+							amount_in
+						} else if v == asset_out {
+							amount_out
```

Comment:
how would you distinguish positive from negative delta in the hook?

Reply:
you have the information which one is asset in and asset out. 

altough  in the adapter, we dont use that. we just provide information of delta_amount and state after the trade. so i belive oracle can work this out? 

it is basically similar to what it done in omnipool adapter, you provide deltas and reserves after trade.  It does not specify which one isnegative